sql/log_event_server.cc

```diff
@@ -8076,44 +8098,20 @@ int Rows_log_event::update_sequence()
 }
 
 
-#endif
-
-
-#if defined(HAVE_REPLICATION)
+void Write_rows_log_event::make_thd_info(thd_info_cache cache)
+{
+  Rows_log_event::make_thd_info(cache, "Write_rows_log_event", {"write_row"});
+}
 
-int
-Write_rows_log_event::do_exec_row(rpl_group_info *rgi)
+int Write_rows_log_event::do_exec_row(rpl_group_info *rgi, thd_info_cache cache)
 {
-  DBUG_ASSERT(m_table != NULL);
-  const char *tmp= thd->get_proc_info();
-  char *message, msg[128];
-  const LEX_CSTRING &table_name= m_table->s->table_name;
-  const char quote_char=
-    get_quote_char_for_identifier(thd, table_name.str, table_name.length);
-  my_snprintf(msg, sizeof msg,
-              "Write_rows_log_event::write_row() on table %c%.*s%c",
-              quote_char, int(table_name.length), table_name.str, quote_char);
-  message= msg;
   int error;
-
-#ifdef WSREP_PROC_INFO
-  my_snprintf(thd->wsrep_info, sizeof(thd->wsrep_info) - 1,
-              "Write_rows_log_event::write_row(%lld) on table %c%.*s%c",
-              (long long) wsrep_thd_trx_seqno(thd), quote_char,
-              int(table_name.length), table_name.str, quote_char);
-  message= thd->wsrep_info;
-#endif /* WSREP_PROC_INFO */
-
-  thd_proc_info(thd, message);
+  #ifdef WSREP_PROC_INFO
+    strmake_buf(thd->wsrep_info, cache[0]);
+  #endif
```

Comment:
Speaking of `thd->wsrep_info`, it’s unfortunate that it has its own dedicated memory where setters still must copy their messages, be it generated or a static constant.
It also leaves it vulnerable to mid-overwrite amalgamation at all times, and the only solution is actually synchronizing.